File: leetcode/longest_substring.cpp
#include <iostream>
#include <string>
#include <unordered_set>

using namespace std;

//finding length of longest substring in a string using sliding window in c++
class Solution {
public:
    int lengthOfLongestSubstring(string s){
        int start = 0;
        int initialstart = 0;
        int end = 0;
        unordered_set<char> uniquechars;
        int maxlen = 0;
        
        while (end<s.length()){
            //not duplicate char
            if (uniquechars.count(s[end])==0) {
                //insert char into set
                uniquechars.insert(s[end]);
                end++;
            }
            
            //duplicate char
            else{
                maxlen = max(maxlen, end-1-start+1);
                initialstart = start;
                end++;
                while (start!=end){
                    uniquechars.erase(s[start]);
                    start++;
                }
                
                //updating start and end to be index immediately after original start
                start = initialstart+1;
                end = start;
            }
        }
        maxlen = max(maxlen, end-1-start+1);
        return maxlen;
        
    }
};
    

int main() {
    string s = "abcadehfera";
    Solution sol;
    cout << sol.lengthOfLongestSubstring(s) << endl;
    return 0;
}

File: leetcode/longest_substring.py
class Solution:
    def __init__(self, s):
        self.start = 0
        self.end = 0
        self.s = s
        self.maxlen = 0
    
    def longest(self):
        unique_chars = []
        while self.end < len(self.s):
            if self.s[self.end] not in unique_chars:
                unique_chars.append(self.s[self.end])
                self.end += 1
                self.maxlen = max(self.maxlen, self.end - self.start)
            else:
                unique_chars.remove(self.s[self.start])
                self.start += 1
        return self.maxlen

s = "abcadehfera"
sol = Solution(s)
print(sol.longest())
